Assertion failed: (total_read == rv_get_reg(rv, rv_reg_a2)), function syscall_read

After commit 56766c0653526c546f00784bbc0cefb800a6f9a2, Doom fails to launch:
```
V_Init: allocate screens.
M_LoadDefaults: Load system defaults.
Assertion failed: (total_read == rv_get_reg(rv, rv_reg_a2)), function syscall_read, file syscall.c, line 290.
/bin/sh: line 1: 12197 Abort trap: 6           ../build/rv32emu doom.elf
make: *** [doom] Error 134
```

Tested on both Ubuntu Linux 20.04 (Intel Xeon E5-2650) and macOS 14.1 (Apple M1).

--- a/src/syscall.c
+++ b/src/syscall.c
@@ -281,13 +281,19 @@ static void syscall_read(riscv_t *rv)
     while (count > PREALLOC_SIZE) {
         size_t r = fread(tmp, 1, PREALLOC_SIZE, handle);
         memory_write(s->mem, buf + total_read, tmp, r);
-        count -= PREALLOC_SIZE;
-        total_read += PREALLOC_SIZE;
+        count -= r;
+        total_read += r;
+        if (r != PREALLOC_SIZE)
+            break;
     }
     size_t r = fread(tmp, 1, count, handle);
     memory_write(s->mem, buf + total_read, tmp, r);
     total_read += r;
-    assert(total_read == rv_get_reg(rv, rv_reg_a2));
+    if (total_read != rv_get_reg(rv, rv_reg_a2) && ferror(handle)) {
+        /* error */
+        rv_set_reg(rv, rv_reg_a0, -1);
+        return;
+    }
     /* success */
     rv_set_reg(rv, rv_reg_a0, total_read);
 }

